sdks/python/apache_beam/runners/worker/sdk_worker.py

```diff
@@ -607,12 +607,16 @@ def _schedule_periodic_shutdown(self):
     # type: () -> None
     def shutdown_inactive_bundle_processors():
       # type: () -> None
+      access_times = []
       with self._lock:
-        for descriptor_id, last_access_time in self.last_access_times.items():
-          if (time.time() - last_access_time >
-              DEFAULT_BUNDLE_PROCESSOR_CACHE_SHUTDOWN_THRESHOLD_S):
-            BundleProcessorCache._shutdown_cached_bundle_processors(
-                self.cached_bundle_processors[descriptor_id])
+        # Shutdown can be expensive, keep out of lock
+        access_times = list(self.last_access_times.items())
+
+      for descriptor_id, last_access_time in access_times:
+        if (time.time() - last_access_time >
```

Comment:
could call time() once outside the loop

I'm not sure if it would be faster to just build up list of descriptor ids under the lock instead of copying the whole list.

Reply:
Yeah, that's probably a good idea - updated